feat(encoder): Added support for mastering display color volume and c…

dkozinski opened the pull request:
…ontent light level metadata

dkozinski committed:
feat(encoder): Added support for mastering display color volume and content light level metadata

Signed-off-by: Dawid Kozinski <d.kozinski@samsung.com>

dkozinski committed:
Refactor: Move HDR metadata preparation out of library

- Responsibility for HDR metadata preparation moved from the library to the application side
- Used the liboapv metadata container API to pass the prepared metadata to the library

Signed-off-by: Dawid Kozinski <d.kozinski@samsung.com>

dkozinski commented:
Currently there are no changes in the encoder implementation regarding support for HDR-related metadata; it does not include handling for master-display and max-cll parameters via the command line. 

PR #168 only adds support for the master-display and max-cll parameters to the reference application (request from library users to add the ability to pass metadata to the APV stream: GitHub issue #167). 

This is solely a change in the reference application to support HDR, not in the encoder.

The user can pass mastering display color volume metadata and content light level information metadata to the reference application. Then the reference application passes this data to the metadata container using the public liboapv API (oapvm_set_all()).

dkozinski committed:
Add metadata serialization with network byte order support

Signed-off-by: Dawid Kozinski <d.kozinski@samsung.com>

dkozinski committed:
Added endianness detection and byte order conversion functions

- Detected platform endianness (Big/Little Endian) in CMake configuration
- Added conditional compilation for endianness-specific byte order conversion in oapv_app_enc.c
- Introduced helper macros (bswap16, bswap32, bswap64) and endian conversion functions (be2ne, le2ne, ne2be, ne2le) in oapv_app_util.h
- Updated metadata serialization functions to use new byte order conversion utilities for consistent cross-platform behavior

Signed-off-by: Dawid Kozinski <d.kozinski@samsung.com>

dkozinski commented:
> @dkozinski The metadata is a binary reprsentation, which is different from C-structure. So, I can say it is not possible to store C-structured data to bitstream directly. Futhermore, the metadata syntaxes use big endian, but mostly computing system supports little endian. And, sometimes, C-compiler tries to change the postion of the structure's members during its optimzation process. I could prepare code patch for resoloving this issue with my idea soon.

@kpchoi 
I've added metadata serialization with network byte order support.

kpchoi committed:
refactored code for clarificatin

Signed-off-by: kp5.choi@samsung.com <kp5.choi@samsung.com>

kpchoi committed:
removed unused code

Signed-off-by: kp5.choi@samsung.com <kp5.choi@samsung.com>

kpchoi committed:
implemented read function for MDCV and CLL

Signed-off-by: kp5.choi@samsung.com <kp5.choi@samsung.com>

dkozinski commented:
@kpchoi I’ll check your latest changes and run the tests as soon as possible this morning

kpchoi committed:
set the payload size

Signed-off-by: kp5.choi@samsung.com <kp5.choi@samsung.com>

dkozinski commented on src/oapv_metadata.c:
```diff
@@ -180,6 +180,97 @@ static void meta_free_md(oapv_md_t *md)
+
+    for(i = 0; i < 3; i++) {
+        t = mdcv->primary_chromaticity_x[i];
+        oapv_assert_rv(t >=0 && t < (2^16), OAPV_ERR_INVALID_ARGUMENT);
```
The operator ^ denotes XOR (a bitwise operation), not exponentiation. In the code, 2^16 is used, resulting in 2 XOR 16 = 18, not 65536. The correct syntax for exponentiation is pow(2, 16)  or manually calculating it as 1 << 16
This causes the condition in an assert statement is not met, for values >= 18 [master-display=G(13250,34500)B(7500,3000)R(34000,16000)WP(15635,16450)L(10000000,1)]
Please change (2^16) to (1<<16)

kpchoi replied:
Mistake, It is changed to 0xFFFF

dkozinski commented on app/oapv_app_enc.c:
```diff
@@ -753,6 +767,98 @@ static int update_param(args_var_t *vars, oapve_param_t *param)
+
+static int update_metadata(args_var_t *vars, oapvm_t mid)
+{
+    int ret, size;
```
ret is not initialized at the start of the function, so if everything succeeds, the function returns an undefined value. Please initialize ret to a default value (0 for success).

kpchoi committed:
fixed wrong constant value

Signed-off-by: kp5.choi@samsung.com <kp5.choi@samsung.com>

kpchoi commented on src/oapv_metadata.c:
```diff
@@ -180,6 +180,97 @@ static void meta_free_md(oapv_md_t *md)
+    for(i = 0; i < 3; i++) {
+        t = mdcv->primary_chromaticity_y[i];
+        oapv_assert_rv(t >=0 && t < (2^16), OAPV_ERR_INVALID_ARGUMENT);
+        oapv_bsw_write(&bs, t, 16);
```
Mistake, It is changed to 0xFFFF

kpchoi commented on src/oapv_metadata.c:
```diff
@@ -180,6 +180,97 @@ static void meta_free_md(oapv_md_t *md)
+        oapv_bsw_write(&bs, t, 16);
+    }
+    t = mdcv->white_point_chromaticity_x;
+    oapv_assert_rv(t >=0 && t < (2^16), OAPV_ERR_INVALID_ARGUMENT);
```
Mistake, It is changed to 0xFFFF

kpchoi commented on src/oapv_metadata.c:
```diff
@@ -180,6 +180,97 @@ static void meta_free_md(oapv_md_t *md)
+    oapv_assert_rv(t >=0 && t < (2^16), OAPV_ERR_INVALID_ARGUMENT);
+    oapv_bsw_write(&bs, t, 16);
+
+    oapv_assert_rv(mdcv->max_mastering_luminance < (2^32), OAPV_ERR_INVALID_ARGUMENT);
```
Mistake, It is changed to 0xFFFFFFFF

kpchoi commented on src/oapv_metadata.c:
```diff
@@ -180,6 +180,97 @@ static void meta_free_md(oapv_md_t *md)
+    tu32 = mdcv->max_mastering_luminance;
+    oapv_bsw_write(&bs, tu32, 32);
+
+    oapv_assert_rv(mdcv->min_mastering_luminance < (2^32), OAPV_ERR_INVALID_ARGUMENT);
```
Mistake, It is changed to 0xFFFFFFFF

kpchoi commented on src/oapv_metadata.c:
```diff
@@ -180,6 +180,97 @@ static void meta_free_md(oapv_md_t *md)
+    oapv_bsw_init(&bs, data, 4, NULL); // CLL payload has 4 bytes
+
+    t = cll->max_cll;
+    oapv_assert_rv(t >=0 && t < (2^16), OAPV_ERR_INVALID_ARGUMENT);
```
Mistake, It is changed to 0xFFFF

kpchoi committed:
fixed bug

Signed-off-by: kp5.choi@samsung.com <kp5.choi@samsung.com>

dkozinski commented on src/oapv_metadata.c:
```diff
@@ -180,6 +180,97 @@ static void meta_free_md(oapv_md_t *md)
+    u32 tu32;
+    oapv_bs_t bs;
+    oapv_bsw_init(&bs, data, 24, NULL); // MDCV payload has 24 bytes
+
```
for(i = 0; i < 3; i++) {
        t = mdcv->primary_chromaticity_x[i];
        oapv_assert_rv(t >=0 && t <= 0xFFFF, OAPV_ERR_INVALID_ARGUMENT);
        oapv_bsw_write(&bs, t, 16);

        t = mdcv->primary_chromaticity_y[i];
        oapv_assert_rv(t >=0 && t <= 0xFFFF, OAPV_ERR_INVALID_ARGUMENT);
        oapv_bsw_write(&bs, t, 16);
    }

dkozinski commented on app/oapv_app_enc.c:
```diff
@@ -753,6 +767,98 @@ static int update_param(args_var_t *vars, oapve_param_t *param)
 }
 
+static int parse_master_display(const char* data_string, oapvm_payload_mdcv_t *mdcv)
+{
```
&mdcv->primary_chromaticity_x[1], &mdcv->primary_chromaticity_y[1], // G
        &mdcv->primary_chromaticity_x[2], &mdcv->primary_chromaticity_y[2], // B
        &mdcv->primary_chromaticity_x[0], &mdcv->primary_chromaticity_y[0], // R

dkozinski reviewed (commented):
some changes are still needed

kpchoi commented:
@dkozinski plesae test the current code again.
And, if you need to modify in application code, you could commit your changes.

dkozinski committed:
Refactor primary chromaticity parsing and metadata writing logic

- Swapped the order of primary chromaticity coordinates in 'parse_master_display' function to ensure correct assignment
- Refactor writing primary chromaticity x and y coordinates in 'oapvm_write_mdcv' function to ensure correct order

Signed-off-by: Dawid Kozinski <d.kozinski@samsung.com>

dkozinski reviewed (commented):
commit 9108e20547ded389eabe00c41dffea8b62d67978 Looks good
